subsys/bluetooth/controller/ll_sw/isoal.c

```diff
@@ -0,0 +1,409 @@
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+	struct isoal_sdu_produced *sdu = &sink->sdu_production.sdu;
+
+	/* Allocate a sdu if the previous was filled (thus sent) */
+	const bool sdu_complete = (sink->sdu_production.sdu_available == 0);
+
+	if (sdu_complete) {
+		/* Allocate new clean sdu buffer */
+		err = sink->session.sdu_alloc(
+			sink,
+			pdu_meta,      /* [in]  PDU origin may determine buffer */
+			&sdu->contents  /* [out] Updated with pointer and size */
+		);
+
+		/* Nothing has been written into buffer yet */
+		sink->sdu_production.sdu_written   = 0;
+		sink->sdu_production.sdu_available = sdu->contents.size;
+		LL_ASSERT(sdu->contents.size > 0);
+
+		/* Remember meta data */
+		sdu->status = pdu_meta->meta->status;
+		sdu->timestamp = pdu_meta->meta->timestamp;
+		/* Get seq number from session counter, and increase sdu counter */
+		sdu->seqn = sink->session.seqn++;
+
+		sdu->origin = pdu_meta->meta->origin;
+		sdu->ctx = pdu_meta->meta->ctx;
+	}
+
+	return err;
+}
+
+static isoal_status_t isoal_rx_try_emit_sdu(struct isoal_sink *sink, bool force_emit)
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+	struct isoal_sdu_produced *sdu = &sink->sdu_production.sdu;
+
+	/* Emit a sdu */
+	const bool sdu_complete = (sink->sdu_production.sdu_available == 0) || force_emit;
+
+	if (sdu_complete) {
+		err = sink->session.sdu_emit(sink, sdu);
+	}
+
+	return err;
+}
+
+static isoal_status_t isoal_rx_append_to_sdu(struct isoal_sink *sink,
+					     const struct isoal_pdu_rx *pdu_meta,
+					     bool is_end_fragment)
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+	const uint8_t *pdu_payload          = pdu_meta->pdu->cis.payload;
+	/* Note length field in same location for both CIS and BIS */
+	isoal_pdu_len_t packet_available    = pdu_meta->pdu->cis.length;
+
+	/* TODO error handling and lost packet */
+
+	/* While there is something left of the packet to consume */
+	while (packet_available > 0) {
+		const isoal_status_t err_alloc = isoal_rx_allocate_sdu(sink, pdu_meta);
+		struct isoal_sdu_produced *sdu = &sink->sdu_production.sdu;
+		uint8_t *sdu_buf = &sdu->contents.codewords[sink->sdu_production.sdu_written];
+
+		err |= err_alloc;
+
+		/*
+		 * For this SDU we can only consume of packet, bounded by:
+		 *   - What can fit in the destination SDU.
+		 *   - What remains of the packet.
+		 */
+		const size_t consume_len = MIN(
+			packet_available,
+			sink->sdu_production.sdu_available
+		);
+
+		LL_ASSERT(sdu_buf);
+		LL_ASSERT(pdu_payload);
+		if (pdu_meta->meta->status == ISOAL_PDU_STATUS_VALID) {
+			memcpy(sdu_buf, pdu_payload + pdu_meta->meta->read_offset, consume_len);
+		}
+		sdu_buf     += consume_len;
+		pdu_payload += consume_len;
+		sink->sdu_production.sdu_written   += consume_len;
+		sink->sdu_production.sdu_available -= consume_len;
+		packet_available                   -= consume_len;
+
+		sdu->status |= pdu_meta->meta->status;
+
+		bool force_emit = (packet_available == 0) && is_end_fragment;
```

Comment:
Not a fan of this. We should just refactor the code so this falls out naturally from the classification.

Reply:
Changed name here